### ISSUE TITLE
Colours aren't parsed by the terminal emulator shown in the Web UI

### ISSUE THREAD

**Issue by SerialVelocity** [drone/drone#939]
Hey,

Colours aren't parsed in the terminal emulator shown in the drone Web UI. You get output such as:
![screen shot 2015-03-25 at 14 44 25](https://cloud.githubusercontent.com/assets/422338/6827104/7b72f178-d2fd-11e4-8b88-4350238990d5.png)

The ```[0;;m``` shouldn't be there.

**Comment by bradrydzewski** [comment]
It does render some ansi terminal characters. You can see an example here:
http://test.drone.io/github.com/drone/drone-cli/master/bfe0db0ae33354e023f793548728bc5643d75968

The implementation is just not 100% compliant. This is the file. Pull requests accepted :)
https://github.com/drone/drone/blob/master/server/app/scripts/line_formatter.js

**Comment by daMupfel** [comment]
@bradrydzewski maybe we can use an external dependency. The are a few ansi to html javascript libraries out there. So it might be a good idea to net reinvent the wheel =)

**Comment by bradrydzewski** [comment]
@daMupfel agreed. I've heard good things about [term.js](https://github.com/chjj/term.js/) but haven't had a chance to try it out yet. Do you have any experience with it?

I was hoping to use something that doesn't require a fixed width / height, so that the build logs could take up as much space as the screen allows.

**Comment by SerialVelocity** [comment]
Maybe trying to refactor something like:
https://github.com/paradoxxxzero/butterfly/blob/master/coffees/term.coffee
may work as well.

Butterfly is really nice and if drone can re-use their terminal, it would be amazing.

**Comment by SerialVelocity** [comment]
Acutally, we probably don't need to support anything that isn't basic terminal codes. We (or really I) really just want 256 ansi colour codes to be supported (and other basic ansi codes)

**Comment by bradrydzewski** [comment]
this one seemed to work well:
https://github.com/rburns/ansi-to-html

**Comment by daMupfel** [comment]
I do not have much experience with ansi encoding.
I looked at the one you mentioned and this one:
https://github.com/drudru/ansi_up

But it does not look difficult to implement. Since we can keep your lineformatter and just use the function from the library inside it.

**Comment by SerialVelocity** [comment]
Yeah, I had a look at those two but neither supports 256 colours. When one is chosen, I could probably try and submit a PR for 256 colour support (I'd start with a bug report though as I don't know javascript or coffeescript too well...)

**Comment by hd-deman** [comment]
@bradrydzewski I found this http://buildkite.github.io/terminal/
It's also support inline-images http://buildkite.github.io/terminal/inline-images/
this can solve issues like #966

also there is a gem for easy show screenshots for test in rails https://github.com/buildkite/capybara-inline-screenshot

**Comment by hd-deman** [comment]
but terminal work on serverside (

**Comment by bradrydzewski** [comment]
0.4.0 is using [ansi-to-html](https://github.com/rburns/ansi-to-html) which seems to be working well

**Comment by renatosilva** [comment]
Please read #1558 more carefully. That issue is not a problem with the ANSI-to-HTML library but *your usage of whatever library is being used*. Maybe you plan outsourcing the web terminal itself to some third-party library, but this is currently [not the case](https://github.com/drone/drone/blob/dfea14c/static/scripts/term_buf.js#L31-L33).

**Comment by bradrydzewski** [comment]
@renatosilva the issue remains unclear. If you have found and issue and know the fix, why not send a pull request?

**Comment by tboerger** [comment]
@renatosilva if you read the source correctly you can see that we are already using an external lib and not our own implementation

**Comment by cdaguerre** [comment]
Sorry if I got something wrong but shouldnt we get colored output in the WebUI in 0.4.0 according to [this comment](https://github.com/drone/drone/issues/939#issuecomment-106003175) by @bradrydzewski ?
I'm using 0.4.0 and don't get any colors.
Is there a tweak or something worth mentioning?
By the way just launched an instance of drone through docker cloud and it took me only an hour to get a build working. This is just insane
Awesome work

**Comment by bradrydzewski** [comment]
@cdaguerre it is important to note that colored output is controlled by the program generating the output. Many programs write colored output conditionally based on the environment. If is very common for programs to disable colored output when the terminal is non-interactive terminals (drone runs non-interactive to prevent the build from hanging at a prompt).

**Comment by cdaguerre** [comment]
@bradrydzewski Thanks a lot, just figured it out ;)

**Comment by jhasse** [comment]
Btw: You can get colored output from CMake using:

```yml
  environment:
    - CLICOLOR_FORCE=1
```

There are other environment variables for other programs (`FORCE_COLOR`, `CLANG_FORCE_COLOR_DIAGNOSTICS`, ...). I'm trying to unify them, see https://bixense.com/clicolors/
